synchronization/futex.rs

```diff
@@ -0,0 +1,524 @@
+            // if the lock is held but not by us
+            if tid != gettid_impl() {
+                // :facepalm: we forked but didn't previously update our tid
+                panic!("Failed on unlock: task {} forked to make task {} but didn't initialize the sync logic.");
+            }
+
+            if value & FUTEX_TID_MASK == 0 {
+                // this futex is already unlocked
+                panic!("Multiple unlock of Mutex {:p} by {}", &self, tid);
+            } else {
+                // this mutex is locked but not by us and we didn't fork
+                panic!(
+                    "Mutex {:p} already locked by {} not {}.",
+                    &self,
+                    value & FUTEX_TID_MASK,
+                    tid
+                );
+            }
+        }
+
+        // if we get here, the futex is locked by us right now
+
+        if (self.0).compare_and_swap(tid, 0, Ordering::SeqCst) != tid {
+            // if this transaction fails, people were waiting on this futex (the value is nonzero)
+            // we have to let the kernel handle the unlock
+            let result = (self.0).unlock_pi();
+            let returncode = result.unwrap_or_else(|r| {
+                // we failed to unlock
+                panic!(
+                    "FUTEX_UNLOCK_PI with (futex={:p}) failed due to {}.",
+                    &self, r
+                );
+            });
+            if returncode != 0 {
+                // we failed to unlock
+                panic!(
+                    "FUTEX_UNLOCK_PI with (futex={:p}) failed, unlock result positive and nonzero",
+                    &self
+                );
+            }
+        } else {
+            // no waiters, we don't have to use the kernel!
+        }
+    }
+
+    fn try_lock(&self) -> bool {
+        // Attempt to do the baseline atomic transaction.
+        let value = (self.0).compare_and_swap(0, gettid(), Ordering::SeqCst);
+
+        // If this is not equal to 0, we failed to lock.
+        return value == 0;
+    }
+}
+
+impl<T: RawFutex> Drop for FutexMutex<T> {
+    fn drop(&mut self) {
+        if likely(!std::thread::panicking()) {
+            // if we're already panicking I guess we don't need to panic again
+            // TODO make this log an error instead
+
+            if unlikely(self.islocked()) {
+                panic!("Trying to drop a locked mutex {:p}", &self)
+            }
+        }
+    }
+}
+
+pub struct FutexCondition(Futex);
+
+impl FutexCondition {
+    fn inner_lock(&self, mutex: &FutexMutex<Futex>) {
+        let mutex_result = mutex.mutex_get(false);
+        if mutex_result == MutexLockResult::Ok {
+            // successfully locked
+            return;
+        } else {
+            panic!("Did not expect an error while locking a condition variable mutex");
+        }
+    }
+
+    pub fn wait(&self, mutex: &FutexMutex<Futex>) {
+        let start = self.0.load(Ordering::SeqCst);
```

Comment:
Is the strict ordering necessary, as opposed to using the std::relaxed equivalent, given that that still guarantees atomicity?

Reply:
We need it to be ordered relative to other processes attempting to wait on the same condvar.

_Maybe_ we'd be safe with Acq/Rel...